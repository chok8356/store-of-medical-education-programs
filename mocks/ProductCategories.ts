import { ICategory } from '../types/ICategory'

export const PRODUCTS_CATEGORIES_COUNT = 5

export const categories: ICategory[] = new Array(PRODUCTS_CATEGORIES_COUNT)
  .fill(null)
  .map((x, i) => {
    const id = i + 1
    return {
      id,
      name: `Категория ${id}`
    }
  })
