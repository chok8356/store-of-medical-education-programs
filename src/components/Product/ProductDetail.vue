<template>
  <b-card
    v-if="product"
    class="overflow-hidden">
    <b-row>
      <b-col md="6">
        <b-card-img :src="product.img" />
      </b-col>
      <b-col md="6">
        <b-card-body>
          <h1 class="mb-4">
            {{ product.name }}
          </h1>

          <b-card-text>
            {{ product.description }}
          </b-card-text>

          <b-badge variant="info">
            {{ category.name }}
          </b-badge>

          <div class="d-block mt-4">
            <b-button
              variant="primary"
              @click="addToCart">
              В корзину
            </b-button>
          </div>
        </b-card-body>
      </b-col>
    </b-row>
  </b-card>
</template>

<script lang="ts">
import { BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop } from 'vue-property-decorator'
import { ICategory } from '../../../types/ICategory'
import { IProduct } from '../../../types/IProduct'

@Component({
  name: 'ProductDetail',
  components: {
    BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg
  }
})
export default class ProductDetail extends Vue {
  @Prop({ type: Object as PropType<IProduct> }) product!: IProduct
  @Prop({ type: Object as PropType<ICategory> }) category!: ICategory

  addToCart() {
    this.$emit('add-to-cart')
  }
}
</script>
