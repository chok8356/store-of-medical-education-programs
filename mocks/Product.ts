import { IProduct } from '../types/IProduct'
import { PRODUCTS_CATEGORIES_COUNT } from './ProductCategories'

const PRODUCTS_COUNT = 100

export const products: IProduct[] = new Array(PRODUCTS_COUNT).fill(null).map((x, i) => {
  const id = i + 1
  return {
    id,
    name: `Продукт ${id}`,
    price: id + PRODUCTS_COUNT,
    // eslint-disable-next-line max-len
    img: 'https://media.istockphoto.com/photos/stack-of-medical-student-textbooks-picture-id921644532?k=20&m=921644532&s=612x612&w=0&h=hezQpZ0uyUmbU4QJePKK5JPNWVXD_UWwWPYSg5Rspds=',
    // eslint-disable-next-line max-len
    description: id % 7 + ' Полностью новые главы по биохимии и клеточной биологии, генетике, нервной системе, костям, мышцам и коже, эндокринной и репродуктивной системам, сердечно-сосудистой системе, почечной системе, диете и питанию.',
    category: id % PRODUCTS_CATEGORIES_COUNT + 1
  }
})
