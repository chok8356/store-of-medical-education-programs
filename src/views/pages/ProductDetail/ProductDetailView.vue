<template>
  <div class="product-detail-view">
    <b-button
      class="mb-4"
      @click="backToStore">
      Назад в магазин
    </b-button>

    <div
      v-loading="isProductLoading"
      class="product-detail-view__loader" />

    <ProductDetail
      v-if="detailProduct && category"
      :category="category"
      :product="detailProduct"
      @add-to-cart="onAddToCart" />
  </div>
</template>

<script lang="ts">
import { BButton } from 'bootstrap-vue'
import { Vue, Component } from 'vue-property-decorator'
import { namespace } from 'vuex-class'
import ProductDetail from '@/components/Product/ProductDetail.vue'
import { loading } from '@/directives'
import { ICategory } from '../../../../types/ICategory'
import { IProduct } from '../../../../types/IProduct'

const ProductStore = namespace('product')
const CartStore = namespace('cart')
const CategoryStore = namespace('category')

@Component({
  name: 'ProductDetailView',
  components: { ProductDetail, BButton },
  directives: {
    loading
  }
})
export default class ProductDetailView extends Vue {
  @ProductStore.State detailProduct: IProduct | null
  @ProductStore.Action getDetailProduct: any

  @CategoryStore.State categories: ICategory[]
  @CategoryStore.Action getCategories: any

  @CartStore.Action addToCart: any

  isProductLoading = true

  async created() {
    if (!this.categories.length) await this.getCategories()
    if (this.productIdFromRoute) await this.getDetailProduct(this.productIdFromRoute)
    this.isProductLoading = false
  }

  get category() {
    return this.categories.find(x => x.id === this.detailProduct?.category)
  }

  get productIdFromRoute() {
    const id = this.$route?.params?.id
    return id ? parseInt(id) : null
  }

  backToStore() {
    this.$router.push({ name: 'ProductsList' })
  }

  onAddToCart() {
    if (this.detailProduct) this.addToCart(this.detailProduct.id)
  }
}
</script>
