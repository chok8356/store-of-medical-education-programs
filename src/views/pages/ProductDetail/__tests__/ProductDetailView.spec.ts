import { createLocalVue, shallowMount } from '@vue/test-utils'

import { BButton } from 'bootstrap-vue'
import Vuex from 'vuex'
import { products } from '@/../mocks/Product'
import { categories } from '@/../mocks/ProductCategories'

import ProductDetail from '@/components/Product/ProductDetail.vue'
import CartStore from '@/store/cart'
import CategoryStore from '@/store/category'

import ProductStore from '@/store/product/product'
import ProductDetailView from '../ProductDetailView.vue'

const product = products[0]
const category = categories[1]

const localVue = createLocalVue()
localVue.use(Vuex)

describe('ProductDetailView', () => {
  let wrapper
  let actions
  let store

  beforeEach(() => {
    actions = {
      getCategories: jest.fn(),
      getDetailProduct: jest.fn(),
      addToCart: jest.fn()
    }
    store = {
      modules: {
        cart: {
          ...CartStore,
          actions: {
            addToCart: actions.addToCart
          }
        },
        product: {
          ...ProductStore,
          state: {
            detailProduct: product
          },
          actions: {
            getDetailProduct: actions.getDetailProduct
          }
        },
        category: {
          ...CategoryStore,
          state: {
            categories: [
              category
            ]
          },
          actions: {
            getCategories: actions.getCategories
          }
        }
      }
    }
    wrapper = shallowMount(ProductDetailView, {
      store: new Vuex.Store(store),
      localVue,
      mocks: {
        $route: {
          params: {
            id: product.id.toString()
          }
        },
        $router: { push: () => {} }
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('render correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('created get categories', () => {
    store.modules.category.state.categories = []
    wrapper = shallowMount(ProductDetailView, {
      store: new Vuex.Store(store),
      localVue
    })
    expect(actions.getCategories).toHaveBeenCalled()
  })

  it('get product data if has id in route', () => {
    expect(actions.getDetailProduct).toHaveBeenCalled()
  })

  it('computed product id from route', () => {
    expect(wrapper.vm.productIdFromRoute).toBe(product.id)
  })

  it('computed category by product category', () => {
    expect(wrapper.vm.category).toBe(category)
  })

  it('back to store', () => {
    const routerPush = jest.spyOn(wrapper.vm.$router, 'push')
    const buttonComponent = wrapper.findComponent(BButton)
    buttonComponent.trigger('click')
    expect(routerPush).toBeCalled()
  })

  it('add to cart', async() => {
    const productDetailComponent = wrapper.findComponent(ProductDetail)
    productDetailComponent.vm.$emit('add-to-cart')
    expect(actions.addToCart).toHaveBeenCalled()
  })
})
