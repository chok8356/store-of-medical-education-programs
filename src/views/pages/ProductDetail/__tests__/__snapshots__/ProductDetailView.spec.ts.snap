// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailView render correctly 1`] = `
<div
  class="product-detail-view"
>
  <b-button-stub
    class="mb-4"
    tag="button"
    target="_self"
    type="button"
    variant="secondary"
  >
    
    Назад в магазин
  
  </b-button-stub>
   
  <div
    class="product-detail-view__loader"
  />
   
  <productdetail-stub
    category="[object Object]"
    product="[object Object]"
  />
</div>
`;
